qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,359 @@
+        """Return whether the plugin supports a toggle for considering
+        directionality of 2-qubit gates as ``natural_direction``.
+
+        Refer to the documentation for :class:`~qiskit.transpiler.passes.UnitarySynthesis`
+        for the possible values and meaning of these values.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_pulse_optimize(self):
+        """Return whether the plugin supports a toggle to optimize pulses
+        during synthesis as ``pulse_optimize``.
+
+        Refer to the documentation for :class:`~qiskit.transpiler.passes.UnitarySynthesis`
+        for the possible values and meaning of these values.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_gate_lengths(self):
+        """Return whether the plugin supports taking ``gate_lengths``
+
+        ``gate_lengths`` will be a dictionary in the form of
+        ``{gate_name: {(qubit_1, qubit_2): length}}``. For example::
+
+            {
+            'sx': {(0,): 0.0006149355812506126, (1,): 0.0006149355812506126},
+            'cx': {(0, 1): 0.012012477900732316, (1, 0): 5.191111111111111e-07}
+            }
+
+        where the ``length`` value is in units of seconds.
+
+        Do note that this dictionary might not be complete or could be empty
+        as it depends on the target backend reporting gate lengths on every
+        gate for each qubit.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_gate_errors(self):
+        """Return whether the plugin supports taking ``gate_errors``
+
+        ``gate_errors`` will be a dictionary in the form of
+        ``{gate_name: {(qubit_1, qubit_2): error}}``. For example::
+
+            {
+            'sx': {(0,): 0.0006149355812506126, (1,): 0.0006149355812506126},
+            'cx': {(0, 1): 0.012012477900732316, (1, 0): 5.191111111111111e-07}
+            }
+
+        Do note that this dictionary might not be complete or could be empty
+        as it depends on the target backend reporting gate errors on every
+        gate for each qubit. The gate error rates reported in ``gate_errors``
+        are provided by the target device ``Backend`` object and the exact
+        meaning might be different depending on the backend.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supported_basis(self):
+        """Returns a dictionary of supported basis for synthesis
+
+        This is expected to return a dictionary where the key is a string
+        basis and the value is a list of gate names that the basis works in.
+        If the synthesis method doesn't support multiple basis this should
+        return ``None``.
+
+        If a dictionary is returned by this method the run kwargs will be
+        passed a parameter ``matched_basis`` which contains a list of the
+        basis strings which match the target basis gate set for the
+        transpilation.
```

Comment:
They can (and probably normally should) but don't have to. This is modeled after https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/synthesis/one_qubit_decompose.py#L42-L55 but abstracting the logic to work for any basis dict that a plugin provide. It's basically just an interface to avoid the boiler plate lookup that many plugins will end up implementing.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/6124/commits/5d8539ac37300473270b470574aa8879c975e2f9